### ISSUE TITLE
Run tasks incrementally after a controlled failure

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#13766]
Currently when a task fails, during the next build it is not allowed to run incrementally. (We report a "change" in its successfulness as the cause for the full rebuild.) This is to avoid jumbled out outputs created due to some unexpected error), but in certain circumstances can cause redundant work to happen.

Example: the developer makes a change, and it gets compiled incrementally, but the compiler detects a simple syntax error, and fails the build. Reading the output the developer fixes the problem, but instead of a quick recompilation of the single changed class, they are punished with a full rebuild.

There are two parts to the problem:

1) There is a difference between a _catastrophic failure_ (i.e. when the outputs are left in an unknown state, because of some unexpected problem) and a _controlled failure_ (where the outputs created are well understood, and we can continue incrementally the next time). Gradle does not currently allow tasks to distinguish between the two kinds of failures. This can be implemented by treating a special exception (e.g. `ControlledFailureException`) differently to other errors thrown by the task.

2) Incremental tasks like `JavaCompile` should be able to produce controlled failures by recording their state in a form that allows them to continue later. This might be something each task needs to do on its own, or we might be able to provide some utilities to make it simpler for certain use cases (e.g. when there is a 1-to-1 association between input and output files).

Also, if we do this, it might make sense to store controlled failures in the build cache, too.

---
cc: @gradle/execution

**Comment by altsvetkov** [comment]
I would happy to see this issue to be resolved! 

I think the importance of this issue is underestimated even for Java (yes, `javac` is quite fast, but errors are common part of edit-test dev cycle and even seconds can add up fast), but for Kotlin having a rebuild after an error was so painful, that we had to implement a workaround by reading output to RAM and restoring it on error. The workaround comes with some non-negligeble cost, but it is much better, than having a full rebuild of a module + (due to implementation of Kotlin IC) a rebuild of all downstream Kotlin tasks.
We have an optimization for the workaround in the works, which for the Kotlin project reduces the cost from 10-20 seconds in some benchmarks to a few seconds. If we could report a controlled failure instead, and drop the workaround completely, we could shave off these few seconds as well (which can be a lot for small incremental builds).

**Comment by altsvetkov** [comment]
As for the problems described in the issue:
1. A special exception would be OK for us.
2. Kotlin IC does not require substantial (or probably any) changes, because it is designed to be able to work after a compile error (and it works great without any workarounds with Intellij internal build tool).

**Comment by sfc-gh-akazantsev** [comment]
This might be a dupe for https://github.com/gradle/gradle/issues/13601#issuecomment-794625449

But this is critical issue for us to onboard gradle. 
* Assumption that developers never write bad/broke code and never break anything is 100% incorrect
* Practice shows that 99% of refactoring are still manual and are driven via compile failures
* Compilation time for real world projects in big enterprise companies can be huge!
* I saw hundreds of monolith projects where compilation time is 30 minutes and more
* My current project compile time is about 10 minutes 
* The only way to be effective is to use incremental compilation.

IntelliJ Idea incremental compiler is mostly instant in all such cases and it does not fall back to full rebuild if error was introduced on prev build.

I will highly appreciate if this issue can be prioritized! Thanks!

**Comment by oehme** [comment]
I'm probably going to look into this after https://github.com/gradle/gradle/pull/16536

**Comment by oehme** [comment]
Just discussed this with @wolfs and we came to a pretty simple solution that I'll put up here for discussion:

When running an incremental task that has previously failed, we'd restore the cached outputs from the last successful build and then compute the input changes against the inputs of the last successful build. That way we make it look like the failure never happened. This would be completely transparent and would require no new build or API. The only requirement would be activating the local build cache.

**Comment by sfc-gh-akazantsev** [comment]
> Just discussed this with @wolfs and we came to a pretty simple solution that I'll put up here for discussion:
> 
> When running an incremental task that has previously failed, we'd restore the cached outputs from the last successful build and then compute the input changes against the inputs of the last successful build. That way we make it look like the failure never happened. This would be completely transparent and would require no new build or API. The only requirement would be activating the local build cache.

Sounds really promising! Thank you guys!

**Comment by sfc-gh-akazantsev** [comment]
Hi Guys, is this going to be fixed? this is really hitting us with build not being incremental.

**Comment by oehme** [comment]
I'm on this right now.

**Comment by oehme** [comment]
@lptr asked me to summarize the current state and the options I considered before deciding on the approach in #17203, so here goes:

Right now, incremental tasks are only incremental when the previous execution was either successful (the obvious case) or if it failed, but the failure did not change any of the outputs. In the latter case, Gradle just pretends that the failure never happened. This works for incremental tasks that do all their validation before deleting anything. Unfortunately, this does not cover most compiler tasks. These delete their output files up-front and then call the compiler. So when the compiler fails on a validation issue, the task fails and the outputs have changed. As a result, the next build will be a full recompile, because Gradle can't tell whether it is safe to rerun incrementally.

The original proposal in this issue was to have a marker exception or similar API that allows tasks to signal to Gradle that the changed outputs are "okay to use on another incremental build". The main benefit of this approach is that it would add almost no runtime overhead and thus make failures just as fast as successes. However, I don't believe we should go this way, as it has several problems:

1. Plugin authors need to adopt this new API. Adopting new Gradle APIs often takes months or years until all important plugins are on board.
2. Plugin authors are humans and thus will make mistakes, meaning they will likely mark unrecoverable failures as recoverable, resulting in wrong results. Or they might be too conservative and mark too many things as "unrecoverable" to be on the safe side, negating any performance benefit
3. It would be particularly challenging when multiple failures happen in a row. E.g. the user deletes `A.java`, the task removes `A.class` but then fails. Then the user reverts `A.java`, but instead makes a breaking change to `B.java`. The build fails again without re-creating `A.class`. What should be the `InputChanges` on the next invocation? `A.java` is unchanged compared to the previous build and also compared to the last successful build. But its class file is missing. We'd either need to provide some kind of `OutputChanges` API or take the whole history of execution into account instead of just looking at the last one like we do now. Both would result in more complexity and would make the system harder to grasp for plugin authors.
4. It is yet another concept that people need to understand in order to write a good Gradle tasks. I'd argue there are already too many of those.

Instead, I'm proposing to do the following: If an incremental task has previously failed **and** changed the outputs, we restore the outputs of the last successful build from the build cache. We then give the task all the input changes since that last successful build. As far as the task is concerned, it will look like the failure never happened. If the outputs haven't changed, we do not load them from the cache, as everything is already there. Same goes when there are any reasons that prevent incrementality. In those cases, the outputs will be fully recreated anyway, so no point in loading from the cache. This approach has several benefits:

1. It will work transparently for all existing incremental tasks. Users will get an immediate benefit when upgrading Gradle
2. No new API and thus no mental overhead for plugin authors
3. Not involving the plugin authors means we avoid a whole class of potential errors
4. API changes usually require a lot of discussion and thus delay delivery of a feature. Not needing an API change means we can deliver this faster.
5. If we later determine that the cache load is still not fast enough for some uses cases, we can always add the "controlled failure" feature on top of this work. I.e. some tasks could use a marker exception and all others would fall back to the cache load behavior instead. I don't think this will be necessary. Loading from the cache is a fast operation and is constantly being optimized  by the Gradle team. I just wanted to point out that the two approaches are not at odds and that the one I'm proposing is the lower hanging fruit.

@lptr @wolfs please let me know whether you agree that this is the way to go. Have a look at #17203 for a rough idea how this can be implemented. If you agree, I will polish that PR up and add the necessary test coverage before asking for a full review.

**Comment by lptr** [comment]
Thanks @oehme for writing this up. I agree that the simplicity of your proposal towards end users and plugin authors has benefits. However, in return for that simplicity it adds complexity on the execution process and the storage of execution state. I'd like to think through if this is a good trade-off or not before committing to making the changes you propose. I hope to find the time for that soon.

**Comment by oehme** [comment]
> However, in return for that simplicity it adds complexity on the execution process and the storage of execution state.

The other proposal (custom exception) would add more complexity (see point 3. in the disadvantages list).

**Comment by oehme** [comment]
@lptr did you have some time to think about this?

**Comment by lptr** [comment]
Sorry, no. There are many competing priorities now and I am not sure we'll get to this in the foreseeable future.

**Comment by lptr** [comment]
Let ma add a bit more context here. We value the work done by contributors highly, and appreciate the features we gain this way, like your improvements around incremental compilation that we've integrated recently. These external contributions have been dominating our workload recently, and we now need to focus on some fundamental infrastructural work before we can get back to them. The plan is to get the integration of external contributions into a more structured form later on, allocating time regularly to attend to them without them monopolizing our attention. Let's revisit this in September, hopefully then we'll be in a better position about this.

**Comment by oehme** [comment]
Okay, thanks for letting me know!

**Comment by lptr** [comment]
We had a good chat about this with @asodja just now that triggered some thoughts on this.

As @oehme points out, the current system of allowing incremental execution after a failure as long as outputs are unchanged works for simpler tasks, but it doesn't work for some compile tasks. Most specifically it affects our incremental compiler implementation in `JavaCompile`. It's also worth noting that `KotlinCompile` does not suffer from this issue, because the implementation of the task restores the original state should a failure happen. What would it take to implement a similar approach for Java, and can this scale to similar tasks?

IIUC, what currently happens in `JavaCompile` is that we delete any `.class` files we expect would be deleted or modified given the incoming changes _before_ we invoke `javac` with the unchanged `.class` files on the classpath. If compilation fails, we are left with missing `.class` files.

However, this doesn't have to be the case. Instead of removing the `.class` files expected to be removed or replaced, we could move them to a temporary location. Likewise, we could generate new `.class` files in another temporary location. Depending on whether or not the compilation is successful, we could move the new or the old files back into the output directory.

The overhead of doing this would be negligible for the successful case, and the benefits for the failed compilation case would be significant. We could do the same for our (still) experimental incremental `GroovyCompile` task, too.

To take a step back, what we are doing here is making the incremental tasks _transactional._ This is an important realization I think.

I'm not sure how many other tasks are there out there that could benefit from a similar treatment, but we could popularize the concept. We could also offer a feature to task implementations to regenerate a subset of the outputs transactionally to make it easier to write such tasks. We could also help plugin authors verify if their tasks are indeed transactional.

(We could even re-run non-incremental tasks transactionally, moving all their outputs to a temporary location, and moving the original outputs back after a failure. So if you make a mistake, then revert it, your next build could be up-to-date. Though this use case is probably already covered sufficiently by the local cache.)

This solution adds no significant overhead for the happy-day case, and requires no change to how tasks are executed (like the introduction of a special exception).

**Comment by lptr** [comment]
We could take this further, and instead of injecting `InputChanges`, we could require that incremental tasks tell us which output files they are about to modify or remove, and we could do the transactional thing automatically for them. They would be given a location to generate new content to that would be moved back to the original output without the task having to do anything.

The same could work for incremental artifact transforms, too.

**Comment by asodja** [comment]
An update: there is no near future plan (in 2022) to provide a general purpose API to run tasks incrementally after a failure. However, we will implement an incremental compilation after a failure for Java and Groovy (FYI Kotlin already supports that). It is planned for the Gradle 7.6, and if you are interested you can follow the PR https://github.com/gradle/gradle/pull/20950.

If there will be any demand for a general purpose API we might implement it, but right now it seems that there is just a demand to implement incremental compilation after a failure.

**Comment by liblit** [comment]
@asodja wrote:
> If there will be any demand for a general purpose API we might implement it, but right now it seems that there is just a demand to implement incremental compilation after a failure.

For what it's worth, I would want a general-purpose API. I have use cases involving custom (not compilation) tasks that operate on many input files. If some input file fails, it would be beneficial to avoid redoing work for other inputs that already succeeded.
